packages/hash/frontend/src/blocks/userBlocks.tsx

```diff
@@ -0,0 +1,158 @@
+import { BlockMetadata } from "blockprotocol";
+import produce from "immer";
+import {
+  createContext,
+  Dispatch,
+  SetStateAction,
+  useContext,
+  useEffect,
+  useMemo,
+  useState,
+} from "react";
+
+import { useCachedDefaultState } from "../components/hooks/useDefaultState";
+
+export type RemoteBlockMetadata = BlockMetadata & {
+  componentId: string;
+  packagePath?: string;
+};
+
+// @todo - remove this and packagePath once componentId starts being generated on the backend
+const getComponentId = (meta: RemoteBlockMetadata) => {
+  if (meta.componentId) {
+    return meta.componentId;
+  }
+
+  if (meta.packagePath) {
+    return `https://blockprotocol.org/blocks/${meta.packagePath}`;
+  }
+
+  throw new Error("Added a block without packagePath or componentId");
+};
+
+type UserBlocks = RemoteBlockMetadata[];
+
+interface UserBlocksContextState {
+  value: UserBlocks;
+  setValue: Dispatch<SetStateAction<UserBlocks>>;
+  blockFetchFailed: boolean;
+}
+
+/** @private enforces use of custom provider */
+const UserBlocksContext = createContext<UserBlocksContextState | null>(null);
+
+const mergeBlocksData = (
+  oldBlocksData: UserBlocks,
+  newBlocksData: UserBlocks,
+): UserBlocks => {
+  return produce(oldBlocksData, (draftUserBlocks) => {
+    for (const latestUserBlock of newBlocksData) {
+      const matchingUserBlockIndex = draftUserBlocks.findIndex(
+        (userBlock) => userBlock.name === latestUserBlock.name,
+      );
+
+      // @todo Remove need for @ts-expect-error here
+      if (matchingUserBlockIndex === -1) {
+        // @ts-expect-error TS warns `Type instantiation is excessively deep` but this isn't a problem here
+        draftUserBlocks.push(latestUserBlock);
+      }
+
+      if (
```

Comment:
I think this ought to be a else if, because `draftUserBlocks[-1]` is never going to exist?

Reply:
This is the case for when the versions of the blocks are different. I guess we can merge the blocks by always using the version from the API over the one present locally?